### PR TITLE
[redux-api-middleware] Improve Dispatch call

### PR DESCRIPTION
**Pull request by artursvonda**
Refactored the dispatched actions to be typed using Discriminated Union.
The promise results are now correctly typed and include payload type without needing explicit typing.

The `RSAARequestAction` changed default Payload and Meta types but the action shouldn't have been used outside the package anyway.
Kept `RSAASuccessAction` and `RSAAFailureAction` with current default Payload and Meta for backwards compatibility and deprecated them in favour of `RSAAResultAction` with stricter default typings.
Removed the `.catch` test since results aren't actually ever thrown from looking at the source code. It's also the reason I unified the success and failure actions since they aren't really that different and discriminated union takes care of differences.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/agraboso/redux-api-middleware#lifecycle
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by iamandrewluca** [comment]
I'll make a review on Monday.

**Review by unrevised6419 — APPROVED** [review]

File: types/redux-api-middleware/index.d.ts [hunk 1]
```diff
@@ -157,8 +156,7 @@ declare module 'redux' {
      * Useful for react-redux or any other library which could use this type.
      */
     interface Dispatch {
-        (action: RSAAAction): Promise<RSAASuccessAction>;
-        (action: RSAAAction): Promise<RSAAFailureAction>;
+        <Payload, Meta>(action: RSAAAction<any, Payload, Meta>): Promise<RSAAResultAction<Payload, Meta>>;
```
[the remark this answers is not in the document]
Comment (iamandrewluca) [reply types/redux-api-middleware/index.d.ts]: @artursvonda add please `State` generic parameter

[the remark this answers is not in the document]
Comment (artursvonda) [reply types/redux-api-middleware/index.d.ts]: Sure! Any reason to have the defaults? They are inferred from the input anyway.

[the remark this answers is not in the document]
Comment (iamandrewluca) [reply types/redux-api-middleware/index.d.ts]: Defaults are if user does not want to use generics.

[the remark this answers is not in the document]
Comment (artursvonda) [reply types/redux-api-middleware/index.d.ts]: Actually in this case the input generic types are inferred and one doesn't need to provide anything.

```
const action = RSAAAction<any, string, number> = { ... };

dispatch(action).then(result => // the result here is inferred correctly and user didn't need to provide any generics
```

By the way, I realised we're not using State in the output so no reason to put it in the generics list.

[the remark this answers is not in the document]
Comment (iamandrewluca) [reply types/redux-api-middleware/index.d.ts]: > Actually in this case the input generic types are inferred and one doesn't need to provide anything.

In the image above it shows that there is an error if you don't provide generics, `1 argument is requried`

> By the way, I realised we're not using State in the output so no reason to put it in the generics list.

It is used inside `RSAAAction` to define type descriptors

[the remark this answers is not in the document]
Comment (artursvonda) [reply types/redux-api-middleware/index.d.ts]: https://www.typescriptlang.org/play/index.html#code/C4TwDgpgBA4hB2EBOBLAxgHgCoD4oF4osBuAKFLQHt4BnYKFeMAV2AC5YFl0M7V4A5nkIByEWQrU6UZkwCGaANYEo2HAApGLdp0SpMuAJQcsBPFtYSqtegDc5AGxWywCxZqatDxKAHpfRAAW0EgQwMxI8FCgkFDBoQw0UFRIoWjADiAM8ABmyKEAJlBySQAGfIwCpUA I think this example illustrates better what I mean and why we don't need default values here.

And the `State` is used inside the action, but it's irrelevant to the output as you can see in the return type `Promise<RSAAResultAction<Payload, Meta>>;`. As it doesn't provide any value, I left it out.

[the remark this answers is not in the document]
Comment (iamandrewluca) [reply types/redux-api-middleware/index.d.ts]: Ok, I agree

**Review by unrevised6419 — APPROVED** [review]

File: types/redux-api-middleware/index.d.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (artursvonda) [reply types/redux-api-middleware/index.d.ts]: Cool! I appreciate you taking your time to give the feedback. Thanks!

**Pull request merged**